package.json

```
@@ -18,36 +18,46 @@
   },
   "files": [
     "dist/**/*",
-    "esm/**/*"
+    "esm/**/*",
+    "*.d.ts"
   ],
   "publishConfig": {
     "access": "public",
+    "main": "./dist/index.js",
+    "types": "./dist/index.d.ts",
     "exports": {
       ".": {
+        "types": "./dist/index.d.ts",
         "import": "./esm/index.mjs",
         "require": "./dist/index.js"
```

> However it is notoriously difficult to fix the "masquerading as CJS" part; I guess this should be fixed with either we integrate [tshy](https://github.com/isaacs/tshy) (which is not PnP ready) or we go ESM only (after the release of Node.js 22).

In my opnion, this may fix our errors.

```suggestion
        "import": {
          "types": "./dist/index.d.ts",
          "default": "./esm/index.mjs"
        },
        "require": {
          "types": "./dist/index.d.ts",
          "default": "./dist/index.js"
        }
```

---

Actually it doesn't, because in order to be a "ESM" declaration, 

- The type's extension should be `.mts`
- And for each type declaration, we have to specify extensions in its type.

```diff
-export * from './array';
+export * from './array/index.mjs';
```

However this is notoriously difficult in current TypeScript environments unless we introduce some good libraries like tshy.

---

then How about using tsup? I think tsup support this quite easily!
https://github.com/suspensive/react/blob/main/packages/react/package.json#L28-L36

---

Next Pull Request for this is also good for me. If you agree tsup way, I'll make new Pull Request for this!

---

Actually tsup does not support our usecase either.

In order to be tree-shaken well, it is necessary for us to use code splitting. However I find that tsup (and its internal implemenation, esbuild) does not support code splitting properly as rollup's `preserveModules`.

---

Maybe we might try just using tsup's declaration generation...